File: src/stores/workspaces.ts
import { invoke } from '@tauri-apps/api/core'
import { createSignal } from 'solid-js'
import { createStore } from 'solid-js/store'

export type Workspace = {
  id: number
  title: string
  description: string
  is_favourite: boolean
  is_selected: boolean
  created_at: Date
  modified_at: Date
  deleted_at: Date | null
}

export const [currentWorkspace, setCurrentWorkspace] = createSignal<Workspace | null>(null)
export const [workspaces, setWorkspaces] = createStore<Workspace[]>([])

export async function fetchCurrentWorkspace() {
  let workspace = currentWorkspace()
  if (!workspace) {
    workspace = workspaces.find((workspace) => workspace.is_selected === true) ?? null
    if (!workspace) {
      workspace = await invoke<Workspace>('fetch_current_workspace')
    }
    setCurrentWorkspace(workspace)
  }
  return workspace
}

export async function addWorkspace(
  workspace: Omit<Workspace, 'id' | 'created_at' | 'modified_at' | 'deleted_at'>
) {
  setWorkspaces((prev) => [
    ...prev,
    { ...workspace, id: prev[prev.length - 1].id + 1 } as Workspace,
  ])

  return await invoke<boolean>('add_workspace', { workspace })
}

export async function fetchAllWorkspaces() {
  const workspaces = await invoke<Workspace[]>('fetch_all_workspaces')
  setCurrentWorkspace(workspaces.find((workspace) => workspace.is_selected === true) ?? null)
  return setWorkspaces(workspaces)
}

export async function updateWorkspace(payload: { id: number; workspace: Partial<Workspace> }) {
  const { id } = payload

  setWorkspaces((prev) => {
    let index = 0
    let item
    for (index; index < prev.length; index++) {
      item = prev[index]
      if (item.id === id) {
        prev[index] = { ...item, ...payload.workspace }
        if (payload.workspace.is_selected === true) {
          setCurrentWorkspace(prev[index])
        }
      }
    }

    return prev.slice()
  })

  return await invoke<boolean>('update_workspace', payload)
}

export async function deleteWorkspace(id: number) {
  setWorkspaces((prev) => prev.filter((workspace) => workspace.id !== id))
  return await invoke<boolean>('delete_workspace', { id })
}

File: tauri/src/commands/workspaces.rs
use serde::{Deserialize, Serialize};
use sqlx::prelude::FromRow;

use crate::db::DatabaseState;

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct Workspace {
    pub id: u32,
    pub title: String,
    pub description: String,
    pub is_favourite: bool,
    pub is_selected: bool,
    pub created_at: String,
    pub modified_at: String,
    pub deleted_at: String,
}

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct WorkspaceInput {
    pub title: String,
    pub description: String,
    pub is_favourite: bool,
    pub is_selected: bool,
}

#[tauri::command(rename_all = "snake_case")]
pub async fn fetch_all_workspaces(
    state: tauri::State<'_, DatabaseState>,
) -> Result<Vec<Workspace>, tauri::Error> {
    let statement = r#"
    SELECT *
    FROM workspaces
    ORDER BY id DESC;
  "#;

    let pool = &state.pool;
    let result = sqlx::query_as::<_, Workspace>(statement)
        .fetch_all(pool)
        .await
        .expect("Fetching all workspaces failed!");

    Ok(result)
}

#[tauri::command(rename_all = "snake_case")]
pub async fn fetch_current_workspace(
    state: tauri::State<'_, DatabaseState>,
) -> Result<Workspace, tauri::Error> {
    let statement = r#"
    SELECT *
    FROM workspaces
    WHERE
      is_selected = 1;
  "#;

    let pool = &state.pool;
    let result = sqlx::query_as::<_, Workspace>(statement)
        .fetch_one(pool)
        .await
        .expect("Fetching all workspaces failed!");

    Ok(result)
}

#[tauri::command(rename_all = "snake_case")]
pub async fn add_workspace(
    state: tauri::State<'_, DatabaseState>,
    workspace: WorkspaceInput,
) -> Result<bool, tauri::Error> {
    let statement = r#"
    INSERT INTO workspaces
    (title, description, is_favourite, is_selected)
    VALUES ($1, $2, $3, $4);
  "#;

    let pool = &state.pool;
    let result = sqlx::query(statement)
        .bind(&workspace.title)
        .bind(&workspace.description)
        .bind(&workspace.is_favourite)
        .bind(&workspace.is_selected)
        .execute(pool)
        .await
        .expect("Adding workspace failed!");

    if result.rows_affected() == 0 {
        return Ok(false);
    }

    Ok(true)
}

#[tauri::command(rename_all = "snake_case")]
pub async fn update_workspace(
    state: tauri::State<'_, DatabaseState>,
    id: u32,
    workspace: WorkspaceInput,
) -> Result<bool, tauri::Error> {
    let statement = r#"
    UPDATE workspaces
    SET title = $1,
        description = $2,
        is_favourite = $3
        is_selected = $4
    WHERE id = $5;
  "#;

    let pool = &state.pool;
    let result = sqlx::query(statement)
        .bind(&workspace.title)
        .bind(&workspace.description)
        .bind(&workspace.is_favourite)
        .bind(&workspace.is_selected)
        .bind(&id)
        .execute(pool)
        .await
        .expect("Updating chronograph failed!");

    if result.rows_affected() == 0 {
        return Ok(false);
    }

    Ok(true)
}

#[tauri::command(rename_all = "snake_case")]
pub async fn delete_workspace(
    state: tauri::State<'_, DatabaseState>,
    id: u32,
) -> Result<bool, tauri::Error> {
    let statement = r#"
     DELETE FROM workspaces
     WHERE id = $1;
    "#;

    let pool = &state.pool;
    let result = sqlx::query(statement)
        .bind(&id)
        .execute(pool)
        .await
        .expect("Deleting workspace failed!");

    if result.rows_affected() == 0 {
        return Ok(false);
    }

    Ok(true)
}
